lib/src/index.rs

```diff
@@ -550,7 +550,40 @@ impl MutableIndex {
     }
 }
 
-impl Index for MutableIndex {
+trait IndexSegment {
+    fn segment_num_parent_commits(&self) -> u32;
+
+    fn segment_num_commits(&self) -> u32;
+
+    fn segment_parent_file(&self) -> Option<&Arc<ReadonlyIndex>>;
+
+    fn segment_name(&self) -> Option<String>;
+
+    fn segment_commit_id_to_pos(&self, commit_id: &CommitId) -> Option<IndexPosition>;
+
+    /// Suppose the given `commit_id` exists, returns the positions of the
+    /// previous and next commit ids in lexicographical order.
+    fn segment_commit_id_to_neighbor_positions(
+        &self,
+        commit_id: &CommitId,
+    ) -> (Option<IndexPosition>, Option<IndexPosition>);
+
+    fn segment_resolve_prefix(&self, prefix: &HexPrefix) -> PrefixResolution<CommitId>;
+
+    fn segment_generation_number(&self, local_pos: u32) -> u32;
+
+    fn segment_commit_id(&self, local_pos: u32) -> CommitId;
+
+    fn segment_change_id(&self, local_pos: u32) -> ChangeId;
+
+    fn segment_num_parents(&self, local_pos: u32) -> u32;
+
+    fn segment_parent_positions(&self, local_pos: u32) -> Vec<IndexPosition>;
+
+    fn segment_entry_by_pos(&self, pos: IndexPosition, local_pos: u32) -> IndexEntry;
+}
+
+impl<T: IndexSegment> Index for T {
```

Comment:
Blanket impl often leads to surprising conflicts, so I would avoid here. It would conflicts with `impl<T: Index> Index for Box<T>` for example.

Maybe `CompositeIndex` could own `T: IndexSegment` instead, but I'm not sure if that's right design.

Reply:
> Blanket impl often leads to surprising conflicts

Is that true even if the blanket implementation is for a private type (like `IndexSegment`)?

I don't feel strongly about this last commit, and I don't mind dropping it.

Reply:
I don't really understand it either. I'll just drop the commit :)